src/psyclone/domain/nemo/transformations/nemo_arrayaccess2loop_trans.py

```diff
@@ -0,0 +1,328 @@
+        loop = NemoLoop.create(loop_variable_symbol, node_copy,
+                               node_copy.copy(), step, [location.detach()])
+        # Replace the original assignment with a loop containing the
+        # modified assignment.
+        loc_parent.children.insert(loc_index, loop)
+
+        if nloops == 0 and not assignment.walk(Range):
+            # There were previously no loops therefore there was not
+            # inlined kernel node and there are no range nodes in the
+            # array reference, so we now need to add an inlined
+            # kernel.
+            CreateNemoKernelTrans().apply(assignment.parent)
+
+    def validate(self, node, options=None):
+        '''Perform various checks to ensure that it is valid to apply the
+        NemoArrayAccess2LoopTrans transformation to the supplied PSyIR Node.
+
+        :param node: the node that is being checked.
+        :type node: :py:class:`psyclone.psyir.nodes.Node`
+        :param options: a dictionary with options for \
+            transformations. No options are used in this \
+            transformation. This is an optional argument that defaults \
+            to None.
+        :type options: dict of string:values or None
+
+        '''
+        # Not a PSyIR node
+        if not isinstance(node, Node):
+            raise TransformationError(
+                "Error in NemoArrayAccess2LoopTrans transformation. The "
+                "supplied node argument should be a PSyIR Node, but found "
+                "'{0}'.".format(type(node).__name__))
+        # Not within an array reference
+        if not node.parent or not isinstance(node.parent, ArrayReference):
+            raise TransformationError(
+                "Error in NemoArrayAccess2LoopTrans transformation. The "
+                "supplied node argument should be within an ArrayReference "
+                "node, but found '{0}'.".format(type(node.parent).__name__))
+        array_ref = node.parent
+        # Array reference not within an assignment
+        if not array_ref.parent or not isinstance(array_ref.parent,
+                                                  Assignment):
+            raise TransformationError(
+                "Error in NemoArrayAccess2LoopTrans transformation. The "
+                "supplied node argument should be within an ArrayReference "
+                "node that is within an Assignment node, but found '{0}'."
+                .format(type(array_ref.parent).__name__))
+        assignment = array_ref.parent
+        # Array reference not on lhs of the assignment
+        if assignment.lhs is not array_ref:
+            raise TransformationError(
+                "Error in NemoArrayAccess2LoopTrans transformation. The "
+                "supplied node argument should be within an ArrayReference "
+                "node that is within the left-hand-side of an Assignment "
+                "node, but it is on the right-hand-side.")
```

Comment:
"but '{0}' is on the RHS of '{1}'".format(my_visitor(node), my_visitor(assignment)) using the trick of supplying a Visitor like you do in adjointificaterisation?

Reply:
I was going to say this should be another PR but I've done as suggested and added an optional writer to the transformation base class constructor so all transformations can now use a writer if they want. I also decided to update the name method in the base class so we don't need to add a name property to a transformation if we simply want to return the class name.